File: LeetCode/0541-Reverse-String-II/541.js
/**
 * @param {string} s
 * @param {number} k
 * @return {string}
 */
var reverseStr = function(s, k) {
    const n = s.length;
    let res = Array.from(s);
    const reverse = (arr, l, r) => {
        while (l < r) {
            [arr[l], arr[r]] = [arr[r], arr[l]];
            l++;
            r--;
        }
    };
    for(let i = 0; i < n; i += 2 * k) {
        reverse(res, i, Math.min(i + k, n) - 1);
        // let l = i - 1, r = i + k > len ? len : i + k;
        // while(++l < --r) [resArr[l], resArr[r]] = [resArr[r], resArr[l]];
    }
    return res.join("");
};

File: LeetCode/0541-Reverse-String-II/541.cpp
/*
LeetCode 541: Reverse String II
Given a string s and an integer k, reverse the first k characters for every 2k
characters counting from the start of the string.

If there are fewer than k characters left, reverse all of them.
If there are less than 2k but greater than or equal to k characters, then reverse the
first k characters and left the other as original.

Constraints:
1 <= s.length <= 10^4
s consists of only lowercase English letters.
1 <= k <= 10^4
*/
#include <algorithm>
#include <iostream>
#include <string.h>
#include <vector>
using namespace std;
/*
思路：这都忘了?orz

*/

class Solution {
public:
	string reverseStr(string s, int k) {
		int n = s.length();
		for (int i = 0; i < n; i += 2 * k)
			reverse(s.begin() + i, s.begin() + min(i + k, n));
		return s;
	}
};
int main() { return 0; }